### ISSUE TITLE
cognito.user.signOut() does not invalidate tokens

### ISSUE THREAD

**Issue by jiachen247** [aws-amplify/amplify-js#3435]
**Describe the bug**
On calling `state.cognito.user.signOut()`, session tokens are just removed localstorage. 
The actual access tokens and refresh tokens are still valid for the lifecycle of the token.

**Expected behavior**
This is a security issue. Best practice dictates session tokens should be invalidated server side on a logout request not just deleted on the client. (OWASP session management)

Source Code found here
https://github.com/aws-amplify/amplify-js/blob/68a5ad2fe2b1d9f03cce80d1bf449e454b621760/packages/aws-amplify-react/src/Auth/SignOut.jsx

**Comment by mlabieniec** [comment]
@jiachen247 if you are using the cognito-identity-js sdk directly, then the `globalSignOut` method will invalidate all sessions (see use case #15 here: https://github.com/aws-amplify/amplify-js/tree/master/packages/amazon-cognito-identity-js). This i believe is what you are looking for to invalidate the server side. The user logout is specifically there as to not invalidate other sessions and just sign out the current local user.

```javascript
cognitoUser.globalSignOut(callback);
```

**Comment by jiachen247** [comment]
@mlabieniec Thanks for your response! yes and no actually. Its close but its still different. global sign out invalidates all open sessions. lets say I log into the webapp and a mobile app both use Cognito for authentication. On global signout both will get logged out. However, signout should just invalidate that particular session on the client side as well as the server side.

**Comment by cbernardes** [comment]
I could not agree more with @jiachen247. I have the same situation of mobile and app usage and the session should be invalidated for each one separately. What actually concerns me more is the fact that I can still use the token after using the "signOut" method of @aws-amplify/auth.

**Comment by jani937** [comment]
Yes, i am facing the same issue. I am using Auth.SignOut({global: ture}) and it is kicking user to login page of that particuler window. If i open the app in another tab in the same window / access the app from other browser or machine they still find a valid user object from id token. I can still get hold of the application. 

I am looking for a solution where if i logout from other tab and come back to another tab and do any operation, it should kick me back to login page.

Any workaround solution for this?

**Comment by ProdigySim** [comment]
Is there some API endpoint we can hit to invalidate a single Refresh token? `Auth.signOut({ global: true })` revokes all refresh tokens, and `Auth.signOut()` revokes no refresh tokens...

There must be some middleground available....

**Comment by cbernardes** [comment]
We are all waiting for it to be fixed.

**Comment by cheeseandcereal** [comment]
Any news on this? It seems ridiculous that you can't invalidate a single set of tokens on the back end.

**Comment by cbernardes** [comment]
@cheeseandcereal funny comment but true. I wrote an article approaching this, so if someone has reached this page, the article may help.

https://medium.com/@cbernardes/serverless-things-i-wish-i-had-known-before-i-started-part-1-aws-cognito-cf5d3a0c3d9d

**Comment by asnar** [comment]
any updates from?

**Comment by joakimDiginex** [comment]
We are struggling hard with this issue and might need to leave Cognito altogether if not solved. As mentioned before, this is a security issue and mentioned in OWASP ASVS etc. Will not pass general security audits.

**Comment by Juank628** [comment]
I'm facing the same problem. It's ridiculous!

**Comment by hckr** [comment]
Is there someone to actually answer these issues/questions? Too frequently I see only tags added and more people writing they have the same problem. No actual support whatsoever and that's sad.

I also need the ability to:
* sign out a single user session (only this one with which the sign out is requested) followed by the token invalidation
* get information (an event?) that the session was invalidated (for example with a global sign out), so I could send the user back to the login screen, as the token is not usable nevertheless, but all Amplify functions don't seem to care, just throw errors.

**Comment by wlarch** [comment]
We encountered the same problem with the AWS Cognito PHP SDK. It seems the documentation is clear for the AdminUserGlobalSignOut function : 

> Signs out users from all devices, as an administrator. It also invalidates all refresh tokens issued to a user. The user's current access and Id tokens remain valid until their expiry. Access and Id tokens expire one hour after they are issued.

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminUserGlobalSignOut.html

A much-needed feature would be to invalidate all tokens on AdminUserGlobalSignOut. Our use case : a user password as been changed by an admin or user, thus the user should be logged out automatically. By the documentation, the user would only be disconnected one hour after...

**Comment by jo-ma** [comment]
> A much-needed feature would be to invalidate all tokens on AdminUserGlobalSignOut. Our use case : a user password as been changed by an admin or user, thus the user should be logged out automatically. By the documentation, the user would only be disconnected one hour after...

By their nature, there is no mechanism to invalidate an access or id token.

The likely solution for your scenario is to track any revoke token events in your app.  Then, as part of your token validation process, you can check to see if you've revoked that specific token and if so not authorize the request.  You could maintain a database table of revoked tokens, cache the values for performance and have a hygiene process to remove expired, revoked tokens.

Your use-case / scenario is different from the original issue.

**Comment by ProdigySim** [comment]
> By their nature, there is no mechanism to invalidate an access or id token.

That's my understanding of JWTs as well. However, I've seen evidence that Cognito tracks individual access tokens.

If you do a global signout, but save your JWT tokens, and then try to hit another Cognito endpoint (like "global signout" again), you'll get a 400 with the message `Access Token has been revoked`. This suggests that Cognito is in fact tracking revocation of individual access tokens in some way.

**Comment by maros96** [comment]
> If you do a global signout, but save your JWT tokens, and then try to hit another Cognito endpoint (like "global signout" again), you'll get a 400 with the message `Access Token has been revoked`. This suggests that Cognito is in fact tracking revocation of individual access tokens in some way.

When you do Global Signout, secret key(or private key and with that also public key) is changed and every token that was created before calling Signout is invalid. So Cognito isn't tracking which token is revoked and which isn't. When key changes, validation of each token fails.

**Comment by soyiatgit** [comment]
Facing the same issue with cognito-identity for JS. `user.signout()` just removes the token from local storage. Anyone who had preserved the id token can still connect to the gateway with it. Any updates or resolution around this ?

**Comment by maros96** [comment]
> Facing the same issue with cognito-identity for JS. `user.signout()` just removes the token from local storage. Anyone who had preserved the id token can still connect to the API gateway with it. Any updates or resolution around this ?

Only way(for now) that is verified from AWS technical support team is explained here:
https://dzone.com/articles/aws-cognito-user-pool-access-token-invalidation-1

So, idea is to user CognitoIdentityServiceProvider. More about all its methods here: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CognitoIdentityServiceProvider.html

CognitoIdentityServiceProvider.globalSingOutUser(accessToken) revokes access token
CognitoIdentityServiceProvider.getUser(accessToken) either return error if token is revoked or users data if token is valid(not expired or revoked)

So, in your case, each time when user sings out(logs out) from application, you should call this globalSingOutUser(accessToken) method. And in case you want to check if user is singed out or not(if token is revoked or not)  you should be checking also access token using this getUser(accessToken) method. That should be done before calling API Gateway. And if you are worried about someone calling API Gateway with just ID token, you can add this check inside for example Lambda(if API Gateway is calling Lambda afterwards). I don't know whole scenario that you are facing with, but you can find some workaround with this.

**Comment by soyiatgit** [comment]
Thanks @maros96  for the detailed explanation. However, when I tried using accessToken as the Authorisation header to connect with API gateway, my API call failed with 401 -Unauthorised. The same call works fine while sending idToken in the Authorisation header. The AWS doc [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-integrate-with-cognito.html) says we can use either of idToken or accessToken. Is that correct. Can you suggest any cause for this behaviour.

**Comment by maros96** [comment]
> Thanks @maros96 for the detailed explanation.
> 
> However, when I tried using accessToken as the Authorisation header to connect with API gateway, my API call failed with 401 -Unauthorised. The same call works fine while sending idToken in the Authorisation header. The AWS doc [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-integrate-with-cognito.html) says we can use either of idToken or accessToken. Is that correct. Can you suggest any cause for this behaviour.

Yes, thing is that Authorizer can depend on ID Token OR Access Token, but implementation of Authorizer is not the same for those 2 cases. You've created your authorizer to use ID Token. If you want to change it to use Access Token, you have to change firs implementation of authorizer and then put accessToken in Authorization header. That is why you get Unauthorized when you pass accessToken - Authorizer expects ID Token instead. On that link you've sent, on the end, you have links for doing everything in steps(obtain permissions, first create Cognito User Pool, ...) and you should read [Integrate a REST API with an Amazon Cognito User Pool](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-enable-cognito-user-pool.html).

So, you can try to implement Authorizer to use Access Token(I've never done that and I am not 100% sure that in case that Access Token is revoked, authorizer will return Unauthorized or will let you use your API, but you should try that). You can try that and after testing that, put a comment here for everyone else, including me, to see how does that work :) 

In case you don't manage to do that, you can still use ID Token for authorizer and call CognitoIdentitiServiceProvider.getUser(accessToken) after calling API Gateway and before doing everything else. Put that check between API Gateway and rest of your application. For example, as I suggested in previous comment, if you call Lambda from your API Gateway, you can first do this check in Lambda and in case access token is revoked, return error status with message "Access Token is revoked". Otherwise you just continue with Lambda execution(which means that access token is valid).

One more thing, in case access token is invalid in any sense(expired, wrong token use, revoked, signature is not the same as decode(hash(header.payload), publicKey) - which means that someone has change something inside that token), this getUser(accessToken) method will return error. If token is valid, it will return some user data. So, you are not just checking if token is revoked, you are also checking whole validity of Access Token.

**Comment by miyasudokoro** [comment]
We still need a way to invalidate an individual refresh token. I don't want to use Global Sign Out because a user might want to be signed in on multiple devices but just want to sign out of one specific device. It would be a poor user experience.

The current system relies solely on each device just "forgetting" the refresh token if a user logs out on that device. This is not a very secure or reliable way to discard refresh tokens because the tokens themselves remain usable. If you do the forgetting wrong, then the user stays logged in without realizing it. API calls using the "logged out" session still return 200, and so forth.

**Comment by maros96** [comment]
There isn't any way to invalidate any token(their purpose is to be stateless and to be valid until they expire). There is either a way (1) to revoke all tokens that are valid and not revoked at that time for one user or (2) to track everything by yourself. Which would mean that you should have a database or any storage system to store tokens that are "invalid"(they are valid but you want them not to be) and then when user uses some token, to check in that database if token is there(it acts as invalid token) or not(acts as valid token). This way would give you what you are asking for.
You have that well described in stackoverflow Invalidate JSON Web Tokens debate. Link is in my first comment here.
And regarding AWS, currently only way is to use CognitoIdentityServiceProvider and globalSignOut method. As I've understood, they are working on improving this, but for now it is like that. Once again, you can get what you want by storing tokens somewhere and checking each token if it is in there or not. But then tokens lose their stateless feature and become dependent on storage state.

**Comment by miyasudokoro** [comment]
I am keeping this thread alive because it is AWS's job to implement this. They are the only ones who can implement a true invalidation measure for refresh tokens because it is possible to call the token refresh API directly if you know the user pool and client id (and client secret if needed). Global sign out is a poor user experience, period. It is a copout to say we should use it as a workaround indefinitely and keep inconveniencing users. It would be dead simple for AWS to put invalidated refresh tokens in a DB belonging to each user pool and check against it for each token refresh request. They just do not care enough to implement it. I first opened a ticket about this more than 2 years ago. All it takes is one general XSS attack script aiming at refresh tokens to gain renewable user sessions for many/most Cognito users, lasting between 1 day and 10 years. Sounds lucrative to me. Only those who keep their user pool etc. totally secret are safe, but you have to have your own API wrapper around login etc. to do that.

**Comment by sammartinez** [comment]
Thank you to everyone with the feedback and concerns presented in this issue. I apologize that we have not made comments on this issue and I do want to state that we are working internally with the AWS Cognito team on this specific issue. Once we have an update from Cognito regarding this issue, we will provide one. Again, thank you for your concern and feedback.

**Comment by hckr** [comment]
@sammartinez Great! When can we expect the update?

**Comment by shivawahi** [comment]
This is a much needed fix. Please provide a fix at the earliest

**Comment by soyiatgit** [comment]
> In case you don't manage to do that, you can still use ID Token for authorizer and call CognitoIdentitiServiceProvider.getUser(accessToken) after calling API Gateway and before doing everything else. Put that check between API Gateway and rest of your application. For example, as I suggested in previous comment, if you call Lambda from your API Gateway, you can first do this check in Lambda and in case access token is revoked, return error status with message "Access Token is revoked". Otherwise you just continue with Lambda execution(which means that access token is valid).

@maros96  here you say, you can still use ID token for authorizer, but the [link](https://dzone.com/articles/aws-cognito-user-pool-access-token-invalidation-1) that you attached says that 
> User hits API Gateway with Authorization header (Access token - JWT)
May be I didn't get the idea behind this approach. Can you try to elaborate it a bit. Many thanks.

**Comment by maros96** [comment]
> > In case you don't manage to do that, you can still use ID Token for authorizer and call CognitoIdentitiServiceProvider.getUser(accessToken) after calling API Gateway and before doing everything else. Put that check between API Gateway and rest of your application. For example, as I suggested in previous comment, if you call Lambda from your API Gateway, you can first do this check in Lambda and in case access token is revoked, return error status with message "Access Token is revoked". Otherwise you just continue with Lambda execution(which means that access token is valid).
> 
> @maros96 here you say, you can still use ID token for authorizer, but the [link](https://dzone.com/articles/aws-cognito-user-pool-access-token-invalidation-1) that you attached says that
> 
> > User hits API Gateway with Authorization header (Access token - JWT)
> 
> May be I didn't get the idea behind this approach. Can you try to elaborate it a bit. Many thanks.

Yes, thing is that you can configure your API Gateway Authorizer to work with either ID Token or Access Token. Depends on your needs. 
I wrote that in some previous comment, it starts with: "Yes, thing is that Authorizer can depend on ID Token OR Access Token, but implementation of Authorizer is not the same for those 2 cases. ".
I suggest you to read the AWS documentation(especially for this case) and all comments here and maybe watch some youtube video to clarify everything :) of course, if something else remains non clarified, just ask

**Comment by ramburg** [comment]
+1 on this one. The token lives only one hour, but immediate invalidation would be the clean way to go @sammartinez

**Comment by IVB107** [comment]
@sammartinez Any updates on this issue?

**Comment by goemon07** [comment]
+1 waiting for some news as well

**Comment by rafalskyi** [comment]
+1 waiting for some news as well

**Comment by muriloSanto** [comment]
+1

**Comment by rsaylor73** [comment]
+1

**Comment by codingyourlife** [comment]
+1

**Comment by kceb** [comment]
Waiting on this as well

**Comment by damjanvucina** [comment]
Waiting

**Comment by jisraeltoledo** [comment]
> > By their nature, there is no mechanism to invalidate an access or id token.
> 
> That's my understanding of JWTs as well. However, I've seen evidence that Cognito tracks individual access tokens.
> 
> If you do a global signout, but save your JWT tokens, and then try to hit another Cognito endpoint (like "global signout" again), you'll get a 400 with the message `Access Token has been revoked`. This suggests that Cognito is in fact tracking revocation of individual access tokens in some way.

This is not correct, after global signout, if you send the idToken to an API using Cognito authorizer, the request is still valid.

**Comment by undefobj** [comment]
This is correct, JWT tokens are signed and this check is valid until they have expired. Invalidation means an out of band disabling of a token before the expiration time has passed even with a valid signature. This feature is not available in Cognito today however we have passed this request to the service team and they are evaluating it in their roadmap. I'll mark this issue as a Feature Request for the time being.

**Comment by cbernardes** [comment]
I wrote an article approaching this, so if someone has reached this page, the article may help.

https://medium.com/javascript-in-plain-english/serverless-things-i-wish-i-had-known-before-i-started-part-1-aws-cognito-cf5d3a0c3d9d

**Comment by dmjesus89** [comment]
I found a way that you dont need call a lambda just decompile the token but I've just implemented by Java

**Comment by dmjesus89** [comment]
I've found this way.

`  JWSObject jwsObject = JWSObject.parse(header);
                if(jwsObject.getState() == com.nimbusds.jose.JWSObject.State.SIGNED) {
                	//signed
                }  if(jwsObject.getState() == null) {
                	//not signed
                }`

**Comment by dmjesus89** [comment]
`  JWSObject jwsObject = JWSObject.parse(header);
                if(jwsObject.getState() == com.nimbusds.jose.JWSObject.State.SIGNED) {
                	//signed
                }  if(jwsObject.getState() == null) {
                	//not signed
                `}``

**Comment by hckr** [comment]
@dmjesus89 How does it relate to this issue?

**Comment by MadisonEhlers-Vertex** [comment]
Looking for updates here as well. It looks like the Global Sign Out is not an option, because the id and access tokens are still valid [for an hour after they were created](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_GlobalSignOut.html) per aws' global sign out documentation. 
> Signs out users from all devices. It also invalidates all refresh tokens issued to a user. The user's current access and Id tokens remain valid until their expiry. Access and Id tokens expire one hour after they are issued.

**Comment by karocksjoelee** [comment]
Waiting this feature + 1

**Comment by daniloalfredo** [comment]
Waiting this feature + 1

**Comment by janpoltan** [comment]
I'm using the user pools hosted ui for oauth flow when integrating 3rd party apps. Even the logout endpoint won't invalidate the current access tokens.

+1 for this feature.

**Comment by amine250** [comment]
+1 waiting

**Comment by owenashurst** [comment]
@sammartinez any update?

**Comment by kennethesguerra** [comment]
Any update on this? Or do you guys have any workaround? Thanks.

**Comment by JorisLimousin** [comment]
+1, it is unbelievable that such a critical feature is still not available, even though it's now been raised publicly almost a year ago, and received a significant amount of interest from your customers.

**Comment by knap1930** [comment]
@undefobj any update, or is there a workaround? I do not think its acceptable to force my users to globally sign out of all apps!

**Comment by ProdigySim** [comment]
Tokens should be invalidated once they are known to be unused.

A user authenticates against my user pool across multiple devices: browsers, mobile phones, public terminals. They get a separate Refresh Token for each device. If they manually click "log out" on one of those devices, the user will drop its reference to the refresh token. The token will be unused.

If an attacker manages to sniff the refresh token from a public machine, our users should be able to click "Log Out" manually to mitigate having their compromised token.

There is a workaround to do this today: It's to use the `signOut({ global: true })` option. This will revoke ALL refresh tokens for the user. This accomplishes the security goals of revoking the unused token, but it simultaneous logs the user out across all devices.

Developers using Cognito are forced to choose between good security and good user experience here.

Is there some reason Cognito is unable to revoke a single refresh token? It sounds like adding a new API to revoke a token could solve this problem, backwards compatibly. I can only imagine there must be some blocking issue in Cognito's end. Perhaps token lifetime for Cognito is handled by some other upstream project that does not give visibility into individual tokens?

**Comment by cjsedwards** [comment]
A related issue (see the closed issue above, I was referred here), if I remove a user from a user group, the JWT they have is still valid and showing that they are a member of that group.  As a result, Amplify/AppSync continue to give them access to data and operations that they should not.  I need some way to invalidate the JWT server side.

**Comment by ancellink** [comment]
Waiting this issue to be resolved.
AWS API needs to be have a single token revoked instead of forcing the global sign out for all the tokens.

**Comment by trusktr** [comment]
Hello you all! Here's a solution: don't use AWS for authentication (obviously they don't care enough to add this small but important feature).

Besides, have you seen AWS's UI and UX in other areas? It is quite bad, and should really make one rethink the choice of using AWS.

**Comment by yepes** [comment]
+1

**Comment by jerry-le** [comment]
Still waiting ...

**Comment by WSorban** [comment]
Subscribing to this issue as well... There must be a way to invalidate a single token on sign out...

**Comment by anhaya** [comment]
Still waiting

**Comment by jacdx** [comment]
I think we need to hit up the Product team about this issue, we’re not
getting a response here. Anyone know of who that would be?

**Comment by taebox** [comment]
I guess I wouldn't hold my breath. Cognito could put a band-aid on this by storing a unique identifier for each token that was 'invalidated'. It would be slightly antithetical to do with JWT as they are designed so that there is no need to have a database filled with sessions. Sessions sound like what you want, you will have full control over whether they are valid or not. There are plenty of solutions for sessions that are well used and tested. I suggest using one of them.

Related: http://cryto.net/~joepie91/blog/2016/06/13/stop-using-jwt-for-sessions/
Also, sorry, that site really is HTTP only.

**Comment by karocksjoelee** [comment]
Just a quick thought of workaround of this, if we write the same logic of sign-in with sign-out, the difference is sign-out doesn't return access token, refresh token, and id token. 
When you sign-in again, the old token will invalid ( that's the same meaning of sign-out ) 

Any thoughts about this? 


I would plan 100% drop off AWS Cognito in the near future of my application.
Still need to figure out putting together better identify solutions

**Comment by muriloSanto** [comment]
A response from AWS Technical Support

```
Hello,

Thank you for contacting AWS Premium Support. I am Ayushi and I will be assisting you today.

I understand you want to know if we are aware of the  issue that`globalSignOut()` or`signOut()` does not immediately invalidate the token of the Cognito user. Therefore, the user would still be able to access the protected resources using existing tokens. Please correct me if I misunderstood.

From the documentation[1] of the `globalSignOut()`, "This method signs out users from all devices. It also invalidates all refresh tokens issued to a user. The user's current access and Id tokens remain valid until their expiry. Access and Id tokens expire one hour after they are issued". 

As the documentation clearly states that after the token is invalidated, the current tokens are still be valid for 1 hour from the time the tokens are issued. This means that When you perform signOut() or `globalSignOut()`, it just clears the tokens from local cache. However, if the tokens stored somewhere else, it is possible that you could still access the protected resources using the tokens until the current token expires. ie, after 60 minutes of token issue time. I understand this causes some issues such as user would still be able to access the protected resources after they are logged out. Unfortunately, at this moment there are no potential workarounds available to get around this behavior.

As one of our member from service team commented on the issue[2], Cognito service team is working on this feature internally. I understand that the lack of the feature is causing inconvenience. I have added your voice to it along with the rest of our customers that are being actively requesting on this. Please be assured that the service team will take the feedback seriously and work towards improving the Cognito service for better customer experience. However, I will not be able to provide any ETA at this moment as a support engineer, I do not have visibility into service team priorities. Meanwhile, You could keep an eye on our what's new blog [3] for information regarding new feature releases.

Please feel free to reach out to us if you have further queries and I'll be happy to assist you on the same.

References
-------------
[1] https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_GlobalSignOut.html

[2] https://github.com/aws-amplify/amplify-js/issues/3435#issuecomment-578206493

[3] https://aws.amazon.com/new/
```

**Comment by muriloneo** [comment]
A reply from AWS Technical Support.

```
Hello,

Thank you for contacting AWS Premium Support. I am Ayushi and I will be assisting you today.

I understand you want to know if we are aware of the  issue that`globalSignOut()` or`signOut()` does not immediately invalidate the token of the Cognito user. Therefore, the user would still be able to access the protected resources using existing tokens. Please correct me if I misunderstood.

From the documentation[1] of the `globalSignOut()`, "This method signs out users from all devices. It also invalidates all refresh tokens issued to a user. The user's current access and Id tokens remain valid until their expiry. Access and Id tokens expire one hour after they are issued". 

As the documentation clearly states that after the token is invalidated, the current tokens are still be valid for 1 hour from the time the tokens are issued. This means that When you perform signOut() or `globalSignOut()`, it just clears the tokens from local cache. However, if the tokens stored somewhere else, it is possible that you could still access the protected resources using the tokens until the current token expires. ie, after 60 minutes of token issue time. I understand this causes some issues such as user would still be able to access the protected resources after they are logged out. Unfortunately, at this moment there are no potential workarounds available to get around this behavior.

As one of our member from service team commented on the issue[2], Cognito service team is working on this feature internally. I understand that the lack of the feature is causing inconvenience. I have added your voice to it along with the rest of our customers that are being actively requesting on this. Please be assured that the service team will take the feedback seriously and work towards improving the Cognito service for better customer experience. However, I will not be able to provide any ETA at this moment as a support engineer, I do not have visibility into service team priorities. Meanwhile, You could keep an eye on our what's new blog [3] for information regarding new feature releases.

Please feel free to reach out to us if you have further queries and I'll be happy to assist you on the same.

References
-------------
[1] https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_GlobalSignOut.html

[2] https://github.com/aws-amplify/amplify-js/issues/3435#issuecomment-578206493

[3] https://aws.amazon.com/new/
```

**Comment by owenashurst** [comment]
> A reply from AWS Technical Support.
> 
> ```
> Hello,
> 
> Thank you for contacting AWS Premium Support. I am Ayushi and I will be assisting you today.
> 
> I understand you want to know if we are aware of the  issue that`globalSignOut()` or`signOut()` does not immediately invalidate the token of the Cognito user. Therefore, the user would still be able to access the protected resources using existing tokens. Please correct me if I misunderstood.
> 
> From the documentation[1] of the `globalSignOut()`, "This method signs out users from all devices. It also invalidates all refresh tokens issued to a user. The user's current access and Id tokens remain valid until their expiry. Access and Id tokens expire one hour after they are issued". 
> 
> As the documentation clearly states that after the token is invalidated, the current tokens are still be valid for 1 hour from the time the tokens are issued. This means that When you perform signOut() or `globalSignOut()`, it just clears the tokens from local cache. However, if the tokens stored somewhere else, it is possible that you could still access the protected resources using the tokens until the current token expires. ie, after 60 minutes of token issue time. I understand this causes some issues such as user would still be able to access the protected resources after they are logged out. Unfortunately, at this moment there are no potential workarounds available to get around this behavior.
> 
> As one of our member from service team commented on the issue[2], Cognito service team is working on this feature internally. I understand that the lack of the feature is causing inconvenience. I have added your voice to it along with the rest of our customers that are being actively requesting on this. Please be assured that the service team will take the feedback seriously and work towards improving the Cognito service for better customer experience. However, I will not be able to provide any ETA at this moment as a support engineer, I do not have visibility into service team priorities. Meanwhile, You could keep an eye on our what's new blog [3] for information regarding new feature releases.
> 
> Please feel free to reach out to us if you have further queries and I'll be happy to assist you on the same.
> 
> References
> -------------
> [1] https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_GlobalSignOut.html
> 
> [2] https://github.com/aws-amplify/amplify-js/issues/3435#issuecomment-578206493
> 
> [3] https://aws.amazon.com/new/
> ```

In that case this thread needs to be closed to stop the spam since it's now been verified that's it's being worked on internally. That work will be prioritized accordingly as you don't know if there's something more important that needs to be worked on first.  You can still use cognito and it doesn't break anything so it's not that big of an issue.

**Comment by ProdigySim** [comment]
> In that case this thread needs to be closed to stop the spam since it's now been verified that's it's being worked on internally.

This repo uses a stale bot to close issues. So, if we stop bumping the issue, it gets closed. I'd rather see the issue stay open personally, but I am not a maintainer.

**Comment by colinfindlay-nz** [comment]
@owenashurst 
Statements about the Cognito team working on issues are also no guarantee that the issue will be resolved in any sort of timeframe, or at all. 
There are _many many_ basic features that aren't supported in Cognito (Removing custom attributes, password expiry etc) - that have "Been raised to the Cognito team and are being worked on internally" that are still undelivered after 4 years of disappointment. Some of them are fundamental security items (like logging out properly)
Based on previous experience - you don't keep this issue open - don't expect it to ever be done.

**Comment by owenashurst** [comment]
Fair enough

**Comment by rvit34** [comment]
What do you think about this workaround? - https://forums.aws.amazon.com/thread.jspa?messageID=879594

**Comment by nihp** [comment]
I have disabled a user in the cognito and the user is already in the app. Here it not shows any error for the user.

User can access the account after he is disabled. 

So If I reload the app it shows the error that is not authenticated from Auth.currentAuthenticatedUser(). Also I can get the currentSession() of the disabled user.


How can I show the user is disabled message immediately the user is disabled from the cognito?

**Comment by SamMade** [comment]
> What do you think about this workaround? - https://forums.aws.amazon.com/thread.jspa?messageID=879594

this process seems to work for me. I can confirm that:

1) Logging out using aws's hosted logout url <https://XXXXXX.auth.us-east-1.amazoncognito.com/logout?client_id=xxxxxxabcxxxxxxc&logout_uri=> revokes the access token (this does not revoke the refresh token)
2) To revoke the refresh token, the only option as of now is globalsignout (unfortunately, this revokes all refresh tokens for that user).

**Comment by aecorredor** [comment]
It’d be great if the Cognito SDK supported the strategy for mobile apps described in the top answer in this SO question: https://stackoverflow.com/questions/26739167/jwt-json-web-token-automatic-prolongation-of-expiration. 

Basically have a way to invalidate a specific refresh token given the device ID where it was generated from. Looking at the current Cognito API it seems like it wouldn’t be that hard to support right? Given that globalSignOut already exists.

**Comment by kimfucious** [comment]
This is a long thread, so sorry if someone has already written about this.

What I wanted to share is the the opposite of this issue is also a problem:  **If a user is programatically added to a Cognito group, the permissions don't take effect until the user logs off/on.**

In brief, if a graphQL type is setup with an @auth directive, like this (in `schema.graphql`):
```js
type User
  @model
  @auth(
    rules: [
      { allow: owner }
      { allow: groups, groups: ["Managers"], operations: [read, update] }
    ]
  )
  @key(name: "byCompany", fields: ["companyId", "id"]) {
...
```
And then the user is added to the "Managers" group via a (node) lambda, like this:
```js
const addToManagerGroup = (userName, userPoolId) => {
  return new Promise((resolve, reject) => {
    try {
      const params = {
        GroupName: "Managers",
        UserPoolId: userPoolId,
        Username: userName
      };
      const data = cognito.adminAddUserToGroup(params).promise();
      resolve(data);
    } catch (error) {
      reject(error);
    }
  });
};
```
If the user in the "Managers" group tries something like the following without logging off after being added to the "Managers" group, the query is empty (unless they're the owner, which they won't be):
```js
let { data } = await API.graphql(
      graphqlOperation(queries.listFormalityUsers, {
        filter: {
          companyId: { eq: companyId }
        },
        limit: 50
      })
```
If there is a way to allow this, without forcing a signOut, I'd love to learn what that is.

**Comment by ProdigySim** [comment]
@kimfucious Have you tried using the [InitiateAuth](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html) endpoint with `REFRESH_TOKEN` auth to generate new Identity & Access Tokens? Groups & other claims should be on the Identity token, which can be refreshed on demand with this endpoint. You may want to open a separate issue if that does not work.

**Comment by kimfucious** [comment]
Thanks for this @ProdigySim 

I've looked at that, and, frankly, I couldn't find any solid docs or examples that enabled me to learn exactly what I should do with `InitateAuth()`.

At the risk of sounding overly whiney, I wish that Amplify would provide an auth method, like `Auth.refreshTokens()`, but I assume there are reasons why they haven't (yet).

Anyhow, while potentially off-topic to this thread, what I would up doing--and it appears to work thus far--is to call the following after any changes are made to a user in Cognito that would require the user to have updated permissions based on those changes:

```js
const cognitoUser = await Auth.currentAuthenticatedUser();
const session = await Auth.currentSession();
const refreshToken = session.getRefreshToken();
cognitoUser.refreshSession(refreshToken, (error, session) => {
  if (error) {
    console.warn(error);
  } else {
    console.log("🙆 Tokens refreshed!", session);
  }
});
```
This prevents me from having to force a signOut.

**Comment by pinpointpanda** [comment]
We're at nearly 9 months since any update from AWS on this (@sammartinez on Jan 24). As others have stated, this is not something that can properly be fixed other than directly by the AWS team. ProdigySim hit the nail on the head further up the thread on this:

> Developers using Cognito are forced to choose between good security and good user experience here.

We're a multi-platform app, and the inability to handle this token revocation on a case-by-case basis is extremely detrimental to our users' experience. Sadly, there are many things missing with Cognito that force us to either offer a poor UX, or compromise on security in someway; the other example we've run across is that there is no endpoint for re-sending a MFA code. Suggestion is simply 'hang on to the username and password and call login again', which just shouldn't be necessary.

Any chance of an update for this? I'm not sensing any commitment to actually fix this issue. This is a commercial product (admittedly with a rather generous free tier currently!), not an open source community library, so I would expect a little more interest from AWS on this... not least because Cognito is one of the 'gateway drugs' into the AWS ecosystem. If there's this little attention paid to basic functionality in Cognito, it worries me the amount of workarounds and step outs we're going to need for basic functionality in other parts of our AWS infrastructure moving forwards.

TLDR; grumble grumble. Any chance of an update? 😄

**Comment by jpignata** [comment]
Thanks, @pinpointpanda! We're very committed to addressing the concerns on this thread. We're currently working behind the scenes on this. Sorry for saying stay tuned, but, please stay tuned! We will provide an update here as soon as we can.

**Comment by pinpointpanda** [comment]
Thanks for the update @jpignata - will keep my 👀 on the thread!

**Comment by MartinsOnuoha** [comment]
Just faced the same issue and was glad to see the last comment from @jpignata here was 19 days ago. Fingers crossed.

**Comment by wshihdehx** [comment]
What happened to this issue? Has it been fixed @jpignata ?

**Comment by CharlieKuharski** [comment]
Any updates?

**Comment by pamelaskaf** [comment]
Any updates?!

**Comment by sonphnt** [comment]
any good news about this?

**Comment by sammartinez** [comment]
Hello all! I wanted to provide an update on this feature. We are actively working on it and will release this feature in the coming months. While I can't give an exact date when this feature will be out, I just want to state that were actively working on it for everyone. Stay tuned for more updates!

**Comment by chatchai-komrangded** [comment]
This's really a must feature, My customer is looking for it, otherwise They may look for alternative.

**Comment by Borduhh** [comment]
I'm not sure if anyone mentioned this already, but it appears you can limit the token expiration now in Cognito all the way down to 5 minutes. Can anyone at Amazon confirm that this would override the default 60 minutes? While it's not a fix for this thread, it should help limit the security vulnerability for sensitive apps.

<img width="583" alt="Screen Shot 2021-03-08 at 8 51 01 AM" src="https://user-images.githubusercontent.com/4759596/110330239-8cba6500-7feb-11eb-9f66-b5bc1179b09c.png">

**Comment by mauriciogzrs** [comment]
Any updates?!

**Comment by stewartcampbell** [comment]
Does this help now? https://aws.amazon.com/about-aws/whats-new/2021/06/amazon-cognito-now-supports-targeted-sign-out-through-refresh-token-revocation/

**Comment by yepes** [comment]
> Does this help now? https://aws.amazon.com/about-aws/whats-new/2021/06/amazon-cognito-now-supports-targeted-sign-out-through-refresh-token-revocation/

Better than nothing, but the thing is how to log out every device right now

**Comment by mlabieniec** [comment]
Hello all, thank you for the continued feedback on this feature request. We've communicated it as a priority to the Cognito team and they have now released [Token Revocation this week](https://aws.amazon.com/about-aws/whats-new/2021/06/amazon-cognito-now-supports-targeted-sign-out-through-refresh-token-revocation/.

We've partnered with them to implement it inside Amplify libraries at launch when calling `Auth.signOut()`. [You can read more about this in the documentation](https://docs.amplify.aws/lib/auth/emailpassword/q/platform/js#sign-out). We are continuing to work with the Cognito team to address any additional feedback. If there are additional feature requests related to this please do open up a new issue.

**Comment by elorzafe** [comment]
> > Does this help now? https://aws.amazon.com/about-aws/whats-new/2021/06/amazon-cognito-now-supports-targeted-sign-out-through-refresh-token-revocation/
> 
> Better than nothing, but the thing is how to log out every device right now

Have you tried global sign-out?, that will invalidate every session for that user

**Comment by sibblegp** [comment]
Is there an answer to this? We want to immediately invalidate all tokens and not wait an hour. We want people signed out of all devices.

**Comment by RubberChickenParadise** [comment]
@sibblegp The main issue is that when App 1 issues a global sign out, unless the App 2 tries to get new tokens from the cognito auth server using the now revoked refresh token, App 2 will never see the global sign out issued from App 1.  You have to build some additional checks on the server or client to check for revoked but still valid due to time tokens.

The whole goal of JWT was to have a decoupled and portable ID token, have an authentication service that issues a token that can be validated with out Service 1 and Service 2 having to call the Authentication server and relying on the preexisting trust relationship secured with cryptographic hashes of the token.  With out tracking the revoked tokens in a way that Service 1 and Service 2 can check, the client will send a "valid" jwt (properly signed and not expired) to the service which will validate the token based on the hash and the expiration time.

When the client goes to exchange the refresh token with cognito for a new access or id token, then the client will get the 401 from cognito because the refresh token is still invalid.

There are a couple ways to handle this:

1. set the access and id token times very low (5 min is the lowest Cognito can go right now).  This will get closer as all the outstanding refresh tokens are now invalid.
2. track revoked access and identity tokens and check that list from the server on every request.  This requires some additional infrastructure since there needs to be a way to track it.  using a cognito trigger to add the tokens to a blacklist dynamo db table is one option.

Its one of the issues with using JWT's, cache invalidation is hard.

**Comment by richardhector** [comment]
Can anyone on this thread confirm that "Amazon Cognito now supports token revocation and Amplify (from version 4.1.0) will revoke Amazon Cognito tokens if the application is online." as per [here](https://docs.amplify.aws/lib/auth/emailpassword/q/platform/js/#sign-out) actually works? After upgrading amplify/auth to

> npm i aws-amplify@4.1.0
> npm i @aws-amplify/auth

there is no invalidation.

Also:
> Is there an answer to this? We want to immediately invalidate all tokens and not wait an hour. We want people signed out of all devices.

Yes this would be extremely helpful from a security perspective.

**Comment by ramburg** [comment]
Can anyone on this thread confirm that "Amazon Cognito now supports token revocation and Amplify (from version 4.1.0) will revoke Amazon Cognito tokens if the application is online." as per [here](https://docs.amplify.aws/lib/auth/emailpassword/q/platform/js/#sign-out) actually works? After upgrading amplify/auth to

> npm i aws-amplify@4.1.0
> npm i @aws-amplify/auth

there is no invalidation.

Also:
> Is there an answer to this? We want to immediately invalidate all tokens and not wait an hour. We want people signed out of all devices.

Yes this would be extremely helpful from a security perspective.

**Comment by ramburg** [comment]
Can anyone on this thread confirm that "Amazon Cognito now supports token revocation and Amplify (from version 4.1.0) will revoke Amazon Cognito tokens if the application is online." as per [here](https://docs.amplify.aws/lib/auth/emailpassword/q/platform/js/#sign-out) actually works? After upgrading amplify/auth to

> npm i aws-amplify@4.1.0
> npm i @aws-amplify/auth

there is no invalidation.

Also:
> Is there an answer to this? We want to immediately invalidate all tokens and not wait an hour. We want people signed out of all devices.

Yes this would be extremely helpful from a security perspective.

**Comment by s1mrankaur** [comment]
@ramburg  came here looking for exactly this.

**Comment by ramburg** [comment]
Hi @jpignata and @sammartinez, it sounds like you are from AWS. As per my post from before, I upgrade all packages to the latest version, but the individual invalidation of a refresh-token doesn't work; there is no client-side call visible, as compare to "global: true", where you can clearly see the call of invalidation in Network commands.

Can you please confirm:
1. whether this feature has been released and in which version?
2. How the invalidation happens if not via a client-side call, visible in the Network tab, where else?
3. If not yet released, when you expect this?

**Comment by ramburg** [comment]
@sethhitch thanks for highlighting! There should be also a hint about it in the Amplify docs [https://docs.amplify.aws/lib/auth/emailpassword/q/platform/js/#re-send-confirmation-code](url) near "Amazon Cognito now supports token revocation and Amplify (from version 4.1.0) will revoke Amazon Cognito tokens if the application is online." in order to ensure this is clear.

**Comment by sethhitch** [comment]
@s1mrankaur 

![Screen Shot 2021-09-13 at 4 17 02 PM](https://user-images.githubusercontent.com/571038/133150558-e5063465-db1e-43f5-a028-610372f4f113.png)
![Screen Shot 2021-09-13 at 4 16 54 PM](https://user-images.githubusercontent.com/571038/133150565-10ad65d4-bdaa-43a7-9c28-4d9c64f3e479.png)

**Comment by s1mrankaur** [comment]
@KristobalJunta  @sethhitch  Thanks. So with https://aws.amazon.com/about-aws/whats-new/2021/06/amazon-cognito-now-supports-targeted-sign-out-through-refresh-token-revocation/ - just the refresh token is revoked, the existing idToken and accessToken continue to be valid?

**Comment by sethhitch** [comment]
@s1mrankaur my experience is consistent with the AWS documentation at https://docs.aws.amazon.com/cognito/latest/developerguide/token-revocation.html

> JWT tokens are self-contained with a signature and expiration time that was assigned when the token was created. Revoked tokens can't be used with any Cognito API calls that require a token. However, Revoked tokens will still be valid if they are verified using any JWT library that verifies the signature and expiration of the token.

So if you try to use an access token associated with the revoked refresh token to make a Cognito API call like `GetUser`, it will fail. However, if you make a REST API call to an API Gateway that uses the built-in Cognito authorizer, and you pass the ID token as a bearer token in the `Authorization` header, it will succeed - the API Gateway authorizer isn't doing revocation checks on the JWT.

**Comment by leonardopaiva** [comment]
I'm asking why I'm implementing my first backend authentication flow... would it be a big security issue to just let the token expire? whereas the token expires in 60 minutes.

Currently what I do is clear the localstorage that holds my token, and every request my endpoint receives depends on receiving the token in the autorization bearer, so without the token in the localstorage the user will have to log in again to use the endpoint again.

Someone mentioned a problem with tabs, the point is that using localStorage is different from using sessionStorage, since localStorage shares the token between tabs, so if I clear localstorage in one tab, the other tab won't be able to use the token anymore because it was cleaned.

But of course it's important to invalidate the token on the endpoint, I'm just asking if there are other situations to consider besides the problem with tabs.

**Comment by revmischa** [comment]
I'm having a problem where a user logs out in a browser extension, invalidating their access token which is shared by the website associated with the extension. The token is successfully revoked, however the Auth API doesn't seem to know what to do when it gets an error from the API.

<img width="1273" alt="Screen Shot 2021-11-29 at 11 46 09 AM" src="https://user-images.githubusercontent.com/245131/143844933-f6b9be5d-7a7b-42d4-89b6-93f27d81e328.png">


Ideally the amplify/cognito/auth library would react to this error automatically by removing the access token from storage and logging the user out, since they clearly aren't logged in anymore. Instead my user is stuck in an in-between state where they're "logged in" but calls to `Auth.currentAuthenticatedUser()` et al fail. User can't do anything and can't log out.

**Comment by LinhTran1** [comment]
> @s1mrankaur my experience is consistent with the AWS documentation at https://docs.aws.amazon.com/cognito/latest/developerguide/token-revocation.html
> 
> > JWT tokens are self-contained with a signature and expiration time that was assigned when the token was created. Revoked tokens can't be used with any Cognito API calls that require a token. However, Revoked tokens will still be valid if they are verified using any JWT library that verifies the signature and expiration of the token.
> 
> So if you try to use an access token associated with the revoked refresh token to make a Cognito API call like `GetUser`, it will fail. However, if you make a REST API call to an API Gateway that uses the built-in Cognito authorizer, and you pass the ID token as a bearer token in the `Authorization` header, it will succeed - the API Gateway authorizer isn't doing revocation checks on the JWT.

This is unbelievable! API Gateway :((

**Comment by mengnans** [comment]
Still facing this issue, do we have a proper solution for it?

**Comment by scorobogaci** [comment]
> I could not agree more with @jiachen247. I have the same situation of mobile and app usage and the session should be invalidated for each one separately. What actually concerns me more is the fact that I can still use the token after using the "signOut" method of @aws-amplify/auth.

why don't you use different app clients, for mobile and web separately ? that's why there are app clients, no?

**Comment by revmischa** [comment]
> why don't you use different app clients, for mobile and web separately ? that's why there are app clients, no?

Because from a user experience point of view it's super lame to ask a user to sign in twice in the same browser, once to your extension and once to your website (in my case)

**Comment by StephenOkeleke97** [comment]
> https://dzone.com/articles/aws-cognito-user-pool-access-token-invalidation-1

This is a solid suggestion. Thanks

**Comment by StephenOkeleke97** [comment]
> why don't you use different app clients, for mobile and web separately ? that's why there are app clients, no?

Solid suggestion.

**Comment by murdaneta** [comment]
With these items you can revoke a single refresh token
`import { CognitoIdentityProvider, RevokeTokenCommand } from "@aws-sdk/client-cognito-identity-provider";`


RevokeTokenCommand
ClientId:The client ID for the token that you want to revoke.
ClientSecret:The secret for the client ID. This is required only if the client ID has a secret.[optional]
Token:The refresh token that you want to revoke.

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_RevokeToken.html

**Here you can see how the implementation is:**
https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-cognito-identity-provider/

**Comment by trunglt-2769** [comment]
@murdaneta 
I cannot send the RevokeTokenCommand, there is an Error: Credential is missing. I'm using userpool of Cognito, the signUp, signIn, refreshToken works fine but error occurs at signOut function
![Screen Shot 2023-06-20 at 19 13 58](https://github.com/aws-amplify/amplify-js/assets/86758445/a4fc6234-1a9e-452e-954d-26deac09d569)

This is providerClient variable:
![Screen Shot 2023-06-20 at 19 16 23](https://github.com/aws-amplify/amplify-js/assets/86758445/5425eaea-fe65-4f3e-83b3-79c737bbf061)

This is signOut function
![Screen Shot 2023-06-20 at 19 16 37](https://github.com/aws-amplify/amplify-js/assets/86758445/20adc18c-773b-47aa-a1ca-5309328ba097)

Do I miss something?

**Comment by essaji** [comment]
From 2019 to 2023 & this is still an issue

**Comment by anaibol** [comment]
use supabase

**Comment by renaldo10** [comment]
**Working workaround:**
GlobalSignOut invalidates `refreshToken` and `accessToken`, but not `idToken` that's why API Gateway calls with cognito authorizer still work. 
I send `idToken` at the Authorization header but in places where i need high security I send `accessToken` in the body/param too, and I do a cognito call of [GetUser](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_GetUser.html) method which takes `accessToken` as param and returns user info, but if the token is invalidated it returns 
`An error occurred (NotAuthorizedException) when calling the GetUser operation: Access Token has been revoked`
Based on this response you can sign out the device directly.

_(even if the Cognito team would invalidate `idToken`, the API Gateway team does simple checks of the sign and expiration date of the token so you would still need the workaround)_

**Comment by ngupta20** [comment]
At this point, AWS should just let us in and do it for them.

**Comment by AleksandarGT** [comment]
Still an issue in 2024. If this issue were a child, I would be sending it to its first day of school soon.

**Comment by BwL1289** [comment]
Is this seriously still not addressed? AWS support said this is "expected behavior." For the life of me, I cannot figure out how this could possibly be as designed. 

Linking to this [comment](https://github.com/aws-amplify/amplify-js/issues/3435#issuecomment-936547922) for my own sanity.

**Comment by rdsedmundo** [comment]
Lots of information have already been said here, I'd just like to add that the [Amplify's documentation makes it seem like the token revocation happens if you call `Auth.signOut({ global: true })`.](https://docs.amplify.aws/react/build-a-backend/auth/concepts/tokens-and-credentials/#:~:text=Token%20revocation%20is,of%20their%20devices.)

But this is not quite I'm looking for, I don't necessarily want to log out the user from their other devices, I just want a SDK call for this endpoint: https://docs.aws.amazon.com/cognito/latest/developerguide/revocation-endpoint.html, [which is a functionality that AWS added only in 2021](https://aws.amazon.com/about-aws/whats-new/2021/06/amazon-cognito-now-supports-targeted-sign-out-through-refresh-token-revocation/), but this issue was closed in 2019, so as far as I can tell it wasn't really implemented, or at least for some reason it was blended with the global sign out unnecessarily—there's also an use case for a separate endpoint or parameter where I can do a local sign out, revoking the local refresh token, leaving the global ones untouched.

@HuiSF @ashika112 @israx it's been years since the last AWS employee last commented here, are you able to shed a light here with a plan for this to be addressed? Should a new issue be created or could this be possibly reopened, or better clarified?

**Comment by jcb-ids** [comment]
Still an issue on 2025, any updates?

**Comment by oceangravity** [comment]
I'm amazed that this is still an issue—unbelievable!

🤌

**Comment by kjkeefe** [comment]
This is still an issue and came up in a security review of my project.

**Comment by wildseansy** [comment]
This also came up in a security review of my app.  This has now been a security gap for 7 years!

**Comment by annjawn** [comment]
2026 and counting.

<img width="351" height="316" alt="Image" src="https://github.com/user-attachments/assets/6860ad3d-c0f9-4369-995d-f1d2d413b04a" />
